module/os/linux/zfs/zfs_ctldir.c

```diff
@@ -999,6 +1009,8 @@ zfsctl_snapshot_unmount(char *snapname, int flags)
 	}
 	rw_exit(&zfs_snapshot_lock);
 
+	exportfs_flush();
```

Comment:
Are there additional consequences from flushing the export table?  Is there a way we can do this only when it's exported?

Reply:
@behlendorf Yes, flushing the export table would flush everything in it, which means it takes time to build up the cache again. By default, the cache expiry time is 120 seconds.

If the snapshot is not mounted, zfsctl_snapshot_unmount() gets no chance to execute exportfs_flush(); I am not sure if there is an easy to check if the snapshot mount point is ever in svc_export and svc_expkey caches, will try to find out.

For regular 'shared' zfs file system, we run something like 'export -u *:<exported_path>' before destroying it. As far as I can tell, export -u ... also flushes everything, it has no ability to flush entries related to the exported path only.

Reply:
That's for the addition information, given the relatively short cache expiry time it doesn't sound like this would be an issue.  The only other question I have is if there's some kernel interface we could use to do this instead of relying on the usermodehelper.  I had a quick look around the nfsd code but didn't find anything suitable.

Reply:
@behlendorf svc_export and svc_expkey caches are in nfsd context, other than parsing /proc/net/rpc/nfsd.fh/content and /proc/net/rpc/nfsd.export/content, I don't an easy way around it. I am not sure if we want to go that way.

Reply:
Yeah, I tend to agree.  But thanks for looking.